distributed/shuffle/shuffle-design.md

```diff
@@ -0,0 +1,139 @@
+    Using an unknown ``shuffle_id`` will first initialize any state needed for that new shuffle.
+    """
+    ...
+```
+
+The `ShuffleExtension` will hold all per-shuffle state and buffers specific to that worker. For example, things like (will vary depending on the stage of the implementation):
+- a buffer of received shards
+- a buffer of outgoing shards (waiting to accumulate enough to be worth sending to a worker)
+- a datastore that transparently spills received shards to disk
+- locks or synchronization mechanisms
+
+The `ShuffleExtension` will also hold any global state shared by all shuffles, such as worker threads or coroutines.
+
+Most of the implementation-specific logic will happen in the `ShuffleExtension`. As we improve performance, add concurrency, etc., this code will change the most (though the interface will likely not).
+
+The `transfer` tasks will pass their input partitions into the `ShuffleExtension`, blocking the task until the extension is ready to receive another input. Internally, the extension will do whatever it needs to do to transfer the data, using worker comms to call the `shuffle_receive` RPC on peer workers. Simultaneously, it'll handle any incoming data from other workers.
+
+### Retries and cancellation
+
+Problems to solve:
+* Causing all tasks in the shuffle to rerun when a worker leaves
+* Cleaning up out-of-band state when a user cancels a shuffle, or it errors
+
+Because most the tasks in the shuffle graph are impure and run for their side effects, restarting an in-progress shuffle requires rerunning _every_ task involved, even ones that appear to have successfully transitioned to `memory` and whose "results" are stored on non-yet-dead workers.
+
+Additionally, cleanly stopping a running shuffle takes more than just releasing the shuffle tasks from memory: since there's out-of-band processing going on, the `ShuffleExtension` has to be informed in some way that it needs to stop doing whatever it's doing in the background, and clear out its buffers. Also, executing tasks may be blocking on the `ShuffleExtension` doing something; without a way to tell the extension to shut down, those tasks might block forever, deadlocking the cluster.
+
+Therefore, we propose adding a `RerunGroup` (`ImpureGroup`? `CoExecutionGroup`? `RestartGroup`? `OutOfBandGroup`? name TBD) structure to the scheduler which intertwines the fates of all tasks within it: if any one task is to be rescheduled (due to its worker leaving), all tasks are restarted; if any one is to be prematurely released (due to cancellation), all are released.
+
+Membership in a `RerunGroup` is implemented via task annotations, where each task gives the name of the `RerunGroup` it belongs to. A task can belong to at most one `RerunGroup`. TBD if we will enforce any structural restrictions on `RerunGroup`s to prevent odd/invalid states from emerging—we probably should, such as not allowing disjoint parts of the graph to be in the same `RerunGroup`, etc.
```

Comment:
> I'm inclined to say this is out of scope and the responsibility of the user.

I think this could create all sorts of headaches for us down the line in the scheduler, where we have to bend ourselves into knots to handle weird cases that are technically possible but nobody should actually do. A more restrictive API up front means less work for us!

Reply:
> Question: What happens if you fuse an annotated blockwise with an unannotated blockwise. The intended behaviour is currently unclear.

There's an open issue around annotations getting lost during optimization https://github.com/dask/dask/issues/7036 but I don't think it fully recognizes the problem (we don't have a clear policy on annotation propagation during optimization). Fixing this would also be necessary work here; I've mentioned it in the doc.